### PR TITLE
Move new search experience to a Beta

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/vector-im/element-meta/issues/139
Fixes https://github.com/vector-im/element-web/issues/20618
Fixes https://github.com/vector-im/element-web/issues/20339
Requires https://github.com/vector-im/element-desktop/pull/314
Requires https://github.com/vector-im/element-web/pull/20895

image
image
image

---
Here's what your changelog entry will look like:

## ✨ Features
 * Move new search experience to a Beta ([\#7718](https://github.com/matrix-org/matrix-react-sdk/pull/7718)). Fixes vector-im/element-meta#139 vector-im/element-web#20618 and vector-im/element-web#20339.

Preview: https://62026ed2e48327bb1a2f7fc7--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

**Review by gsouquet — APPROVED** [review]

**Comment by niquewoodhouse** [comment]
> I think this still needs a toast prompt

https://user-images.githubusercontent.com/1587166/152496938-e0aa391c-b271-4678-ae60-2555bc653a3d.mov

### Copy:
- New search beta available
- We're testing a new search to make finding what you want quicker.
- Dismiss / Learn more

**Review by kittykat — CHANGES_REQUESTED** [review]
* Could you please add a delay for the toast? - I just clicked through the three on autopilot without reading any of them.

* There's a black border around the gif in the settings dialog:
Screenshot from 2022-02-04 13-06-22

* Copy:
  * I kinda feel like *"The" new search experience* sounds a bit more dramatic and exciting
  * Add comma after "new": "A new, quick way…"
  * Maybe "This feature is a work in progress, we'd love to hear your feedback."
  * I'm not really sure what's going on here or what that's trying to tell me?
    >Feedback
    > Search and click on feedback"

* Labs dialog spacing: could you verify that it matches what's expected based on what we did in the sidebar? Probably need design input, possibly from Jano.

* I think we should only be asking for feedback once, and maybe need to review button design? They're very different from each other atm:
  Screenshot from 2022-02-04 13-12-38

**Comment by t3chguy** [comment]
> Could you please add a delay for the toast? - I just clicked through the three on autopilot without reading any of them.

There is a delay of 5 minutes, think you confused it with the old IA toast which is being removed in another PR

> There's a black border around the gif in the settings dialog:

The black border on the left is in the gif itself, as for the rounded corners I'll match them in CSS. @niquewoodhouse <---

> Copy

@niquewoodhouse 

> Labs dialog spacing: could you verify that it matches what's expected based on what we did in the sidebar? Probably need design input, possibly from Jano.

Not sure what you mean by "what we did in the sidebar" here, could you elaborate @kittykat 

> I think we should only be asking for feedback once, and maybe need to review button design? They're very different from each other atm:

@niquewoodhouse

**Review by niquewoodhouse — CHANGES_REQUESTED** [review]
Few small things, nothing major.

**Feedback form**

Think we need to change the copy a bit, as it's referring to a WIP name for it.

- [x] Title: ~~Spotlight search feedback~~ Search beta feedback
- [x] Body: ~~Thank you for trying Spotlight search. Your feedback will help inform the next versions.~~ Thank you for trying the beta, please go into as much detail as you can so we can improve it.  Your platform and username will be noted to help us use your feedback as much as we can.


------

**Search view**

- [x] Beta badge `beta` should be semibold
- [ ] Change copy use.... to scroll ~~results~~ (I don't think the word results is necessary).

**Labs advert**

- [ ] Can a css border on the gif hide the black border in my gif?  If not, I can make new one.
- [ ] There's something a bit confusing about the instant 'join/leave' experience.  Can we use a spinner in the button, or a disabled loading state for 2 seconds between them, or something like this?

**Comment by t3chguy** [comment]
> Can a css border on the gif hide the black border in my gif?

only the one on the corners, not the one on the left, hiding that would be specific to this beta whereas all the beta stuff is very generic so lets avoid that

> Can we use a spinner in the button, or a disabled loading state for 2 seconds between them, or something like this?

I think introducing delays just feels very weird but yeah we can do

**Comment by kittykat** [comment]
> > Labs dialog spacing: could you verify that it matches what's expected based on what we did in the sidebar? Probably need design input, possibly from Jano.
> 
> Not sure what you mean by "what we did in the sidebar" here, could you elaborate @kittykat

@t3chguy The text in the Labs tab starts higher than it probably should:

settings-2

**Comment by t3chguy** [comment]
> The text in the Labs tab starts higher than it probably should:

Alright, though that isn't due to any changes in this PR

**Comment by kittykat** [comment]
> > The text in the Labs tab starts higher than it probably should:
> 
> Alright, though that isn't due to any changes in this PR

Indeed, but I feel it would make sense to fix it for a better user experience for joining the Beta

**Comment by niquewoodhouse** [comment]
> I think we should only be asking for feedback once, and maybe need to review the button design? They're very different from each other atm

I disagree @kittykat.  I think we should be asking for feedback permanently in the dialog.  People search for different things at different times, so they'll find things frustrating/hard when using it one time, then raise different points the next time.  After using it for days you'll be able to provide more feedback than just using it once.  The feedback button looks very muted because we don't want to distract people from their search flow with primary buttons in a footer.   With just asking once I think we'll miss the majority of useful feedback, as search is such a repeated use flow.   

Copy:

@kittykat based on your feedback is this agreeable to you?  Please feel free to amend as you see fit.

Title: The new search `beta`<sup>1</sup>
Body: A new, quick way to search spaces and rooms you're in.  This feature is a work in progress, we'd love to hear your feedback.  
Footer:  **How can I feedback?**  To feedback, join the beta, start a search and click on feedback.  Or alternatively, click here.
**How can I leave the beta?**  To leave, just return to this page or tap on the beta badge when you search.

<sup>1 </sup>  I removed `experience` because it might make the title too long, and make it awkwardly spread to two lines on a maximized window on a normal laptop.  

Thanks for the help

**Comment by t3chguy** [comment]
> Indeed, but I feel it would make sense to fix it for a better user experience for joining the Beta

They start at the same height, but due to different font sizes they also have different line-heights, the Sidebar one sporting a line-height of 2.4rem vs a `normal` line height on the Labs tab. So unless we conform all tabs to having a similar hierarchical structure (having a subheading with exact same font style as the very first thing) this misalignment will persist

**Comment by kittykat** [comment]
> > I think we should only be asking for feedback once, and maybe need to review the button design? They're very different from each other atm
> 
> I disagree @kittykat. I think we should be asking for feedback permanently in the dialog. People search for different things at different times, so they'll find things frustrating/hard when using it one time, then raise different points the next time. After using it for days you'll be able to provide more feedback than just using it once. The feedback button looks very muted because we don't want to distract people from their search flow with primary buttons in a footer. With just asking once I think we'll miss the majority of useful feedback, as search is such a repeated use flow.

Yes permanently, but we have two links for it which are right next to each other. I think we only need one link to "feedback" rather than two :+1: 

> Copy:
> 
> @kittykat based on your feedback is this agreeable to you? Please feel free to amend as you see fit.
> 
> Title: The new search `beta`1 Body: A new, quick way to search spaces and rooms you're in. This feature is a work in progress, we'd love to hear your feedback. Footer: **How can I feedback?** To feedback, join the beta, start a search and click on feedback. Or alternatively, click here. **How can I leave the beta?** To leave, just return to this page or tap on the beta badge when you search.

Much better, thank you! The only bit the writer in me rebels against is that "feedback" is a noun, not a verb. I'd be happier if at least the title could be "How can I give feedback?"

**Comment by niquewoodhouse** [comment]
> Yes permanently, but we have two links for it which are right next to each other. I think we only need one link to "feedback" rather than two 👍

Ah, sorry, I completely misunderstood!  I'd personally leave it at two links to maximize the chance of feedback.  I do appreciate it's a bit of a lazy design but we want as much feedback as we can get.  I think for a beta thing that's more important than a cleaner UI.

If we remove the text link: people might read the copy, see the word feedback, and then click it > nothing happens, they don't look for the button or find it.  If we remove the button: people have to read an entire sentence to find the option to feedback - when in the context of looking at a results page, you're more interested in scanning text than reading whole sentences.

@t3chguy I tweaked the copy below based on the above input.

Title: The new search beta1
Body: A new, quick way to search spaces and rooms you're in. This feature is a work in progress, we'd love to hear your feedback.
Footer: How can I **give** feedback? To feedback, join the beta, start a search and click on feedback. Or alternatively, click here.
How can I leave the beta? To leave, just return to this page or tap on the beta badge when you search.

File: src/settings/Settings.tsx [hunk 1]
```diff
@@ -336,8 +337,23 @@ export const SETTINGS: {[setting: string]: ISetting} = {
...
+        default: false,
+        betaInfo: {
+            title: _td("The new search beta"),
+            caption: () => <>
+                <p>{ _t("A new, quick way to search spaces and rooms you're in.") }</p>
+                <p>{ _t("This feature is a work in progress, we'd love to hear your feedback.") }</p>
+            </>,
+            disclaimer: () => <>
+                <h4>{ _t("How can I give feedback?") }</h4>
+                <p>{ _t("To feedback, join the beta, start a search and click on feedback. Or alternatively, click here.") }</p>
```
Comment (gsouquet) [inline src/settings/Settings.tsx:350]: Should this contain a link of some sort ? the copy says "click here" ?

**Review by gsouquet — APPROVED** [review]
Code looks good to me!

File: src/settings/Settings.tsx (hunk 1 above)
Comment (t3chguy) [reply src/settings/Settings.tsx:350]: Yup, tweak incoming (copy and some other bits)

**Review by gsouquet — APPROVED** [review]

**Review by niquewoodhouse — CHANGES_REQUESTED** [review]
Looking really close, thanks!

A couple of small things.

**On the search button**

- [ ] Align search button copy and icon

I think it's because of the position of the search icon svg within it's artboard/viewbox but the icon and word search don't seem aligned.  I added a margin-bottom 2px to the svg and they seemed better.

image

image

**Join beta in labs**

- [ ] Amend copy for clarity pre-join

Before you join the "how can I give feedback" section doesn't make sense, just needs a minor change in the last sentence.  From `Or alternatively, click the feedback button above` to `Or, once you've joined the beta, click the feedback button above`

**Review by niquewoodhouse — APPROVED** [review]
🥳

**Review by nadonomy — CHANGES_REQUESTED** [review]
This is looking/feeling great! I have a few nits I'd like to raise:

### Recommendations
_Things I feel very strongly on, which I'd like to either address or discuss before approving._
- [ ] In the screenshot below, the micro copy details "click the feedback button above" while a feedback button above doesn't exist. We should either adjust the copy to be more generic, or not show this sentence while a button isn't being displayed.
Screenshot 2022-02-07 at 17 01 24

- [ ] In the screenshot above, we use the verb 'tap' in the last paragraph. I think we should always use 'click' consistently on Web/Desktop, which would also make it consistent with where we use 'click' in this same screen.

- [ ] In the screenshot below, the copy is confusing as (1) it discusses the Spaces beta & (2) uses the word tap, again.

  
Screenshot 2022-02-07 at 17 01 58


### Suggestions
- [ ] I like the new labs icon in the toast! But, I find the inconsistency between this icon and the flag icon in Settings strange. Wdyt to using the test tube in both places?
- [ ] Nit: In the screenshot below, 'Feedback' [isn't a verb](https://ell.stackexchange.com/questions/58023/does-it-sound-ok-to-use-feedback-as-a-verb) so we should -> to 'give feedback' or something else instead.

Otherwise looking great! Really excited to get more feedback in the wild!

**Comment by t3chguy** [comment]
>  I like the new labs icon in the toast! But, I find the inconsistency between this icon and the flag icon in Settings strange. Wdyt to using the test tube in both places?

We used it for the IA toast too, I have no love for the flag icon

**Comment by niquewoodhouse** [comment]
> We used it for the IA toast too, I have no love for the flag icon

>  I like the new labs icon in the toast! But, I find the inconsistency between this icon and the flag icon in Settings strange. Wdyt to using the test tube in both places?

I think it's a really good suggestion to use the same icon in both places.

**Comment by t3chguy** [comment]
Awaiting re-authored flask icon from @niquewoodhouse to better match tab icons

**Comment by t3chguy** [comment]
image

**Review by nadonomy — APPROVED** [review]
Approving for expediency to ease merging, but @t3chguy is it trivial to vertically align this BETA label?

Screenshot 2022-02-08 at 13 03 40

Screenshot 2022-02-08 at 13 03 13

**Comment by t3chguy** [comment]
> but @t3chguy is it trivial to vertically align this BETA label?

Yup

**Pull request merged**